docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,274 @@
+  but for service location and configuration as well.
+  The requirement to synchronize framework startup for
+  the purpose of securely initializing Consul means
+  that a non-Consul mechanism must be used
+  to stage-gate EdgeX initialization.
+
+This last point is the primary motivator of this ADR.
+
+## Decision
+
+### Stage-gate mechanism
+
+The stage-gate mechanism must work in the following environments:
+
+* docker-compose in Linux on a single node
+* docker-compose in Microsoft* Windows* on a single node
+* docker-compose in Apple* MacOS* on a single node
+
+Startup sequencing will be driven by two primary mechanisms:
+
+1. Use of entrypoint scripts to:
+
+   - Block on stage-gate and service dependencies
+   - Perform first-boot initialization phase activities as noted in [Context](#context)
+
+   The bootstrap container will inject entrypoint scripts into
+   the other containers in the case where EdgeX is directly consuming
+   an upstream container.  Docker will automatically retry
+   restarting containers if its entrypoint script is missing.
+
+2. Use of open TCP sockets as semaphores to gate startup sequencing
+
+   Use of TCP sockets for startup sequencing is commonly used in Docker environments.
+   Due to its popularlity, there are several existing tools for this, including
+   [wait-for-it](https://github.com/vishnubob/wait-for-it),
+   [dockerize](https://github.com/jwilder/dockerize), and
+   [wait-for](https://github.com/Eficode/wait-for).
+   The TCP mechanism is portable across platforms
+   and will work in distributed multi-node scenarios.
+
+At least three new ports will be added to EdgeX for sequencing purposes:
+
+* `bootstrap` port. This port will be opened once first-time initialization has been completed.
+* `tokens_ready` port. This port signals that secret-store tokens have been provisioned and are valid.
+* `ready_to_run` port. This port will be opened once stateful services have completed initialization
+  and it is safe for the majority of EdgeX core services to start.
+
+The stateless EdgeX services should block on `ready_to_run` port.
+
+### Docker-specific service changes
+
+#### "As-is" startup flow
+
+The following diagram shows the "as-is" startup flow.
+
+![AS-IS](./bootstrap-as-is.jpg)
+
+There are several components being removed via activity unrelated with this ADR.
+These proposed edits are shown to reduce clutter in the TO-BE diagram.
+* secrets-setup is being eliminated through a separate ADR to eliminate TLS for single-node usage.
+* kong-migrations is being combined with the kong service via an entrypoint script.
+* bootstrap-redis will be incorporated into the Redis entrypoint script to set the Redis password before Redis starts to fix the time delay before a Redis password is set.
+
+#### "To-be" startup flow
+
+The following diagram shows the "to-be" startup flow.
+Note that the bootstrap flows are always processed,
+but can be short-circuited.
+
+![TO-BE](./bootstrap-to-be.jpg)
```

Comment:
The `tokens_ready` port is not shown. I assume vault worker is responsible for this?

Reply:
Yes.

Reply:
Diagram updated.